### PR TITLE
ES.46 Issue 1797 - narrowing to bool

### PR DESCRIPTION
**Pull request by bgloyer**
Added note for narrowing to bool. See  #1797.

Also qualified gsl::narrow

### CODE REVIEW

**Review by jwakely — APPROVED** [review]
I think this note makes sense. There is no problem with "narrowing" here, because there is no data loss in a Boolean test. All non-null pointers are "true", but it's silly to say there is data loss when testing them. Similarly, integers `1` and `100` are both "true" but there is no data loss by testing that. The difference between different non-zero values is not significant here.

**Comment by shaneasd** [comment]
I think this should be coupled with a modification of the note on ES.87 http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es87-dont-add-redundant--or--to-conditions which explicitly contradicts it.

Separately to that, It's not obvious to me why ```bool b = ptr; // bad: narrowing``` is bad when compared to ```if (ptr) do_something(*ptr);```. If ```if (myint)``` should be interpreted as ```if (myint != 0)``` then shouldn't ```bool b = myint``` be interpreted as ```bool b = myint != 0```?

**Comment by jwakely** [comment]
I see a fundamental difference between creating a new variable that holds a narrowed value, and testing a condition.

**Comment by MikeGitb** [comment]
Another fundamental difference is that 

    bool b  = <pointer expression >

is much more likely a typo/error (e.g. the author expected that a function returns a boolean when in reality it returns a pointer) than `if(ptr)` .

**Comment by bgloyer** [comment]
> I think this should be coupled with a modification of the note on ES.87 http://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#es87-dont-add-redundant--or--to-conditions which explicitly contradicts it.
> 
> Separately to that, It's not obvious to me why `bool b = ptr; // bad: narrowing` is bad when compared to `if (ptr) do_something(*ptr);`. If `if (myint)` should be interpreted as `if (myint != 0)` then shouldn't `bool b = myint` be interpreted as `bool b = myint != 0`?

Yea, it is somewhat inconsistent.  I wouldn't mind allowing the assignment to ````bool```` as well for consistency.  It is less dangerous than other narrowing because all possible floating and integral values map unambiguously to bool values unlike types like ````short```` which have problems with large values and NaN.

**Review by ChrisGuzak — APPROVED** [review]

File: CppCoreGuidelines.md [hunk 1]
```diff
@@ -11704,17 +11704,24 @@ A key example is basic narrowing:
...
-    u = d;                          // BAD
-    u = narrow_cast<unsigned>(d);   // OK (you asked for it): u becomes 4294967289
-    u = narrow<unsigned>(d);        // OK: throws narrowing_error
+    u = d;                               // bad: narrowing
+    u = gsl::narrow_cast<unsigned>(d);   // OK (you asked for it): u becomes 4294967289
+    u = gsl::narrow<unsigned>(d);        // OK: throws narrowing_error
+
+##### Note
+
+This rule does not apply to values that are used as a condition:
```
Comment (beinhaerter) [inline CppCoreGuidelines.md:11721]: jwakely noted in https://github.com/isocpp/CppCoreGuidelines/issues/1797#issuecomment-923018795 that the exception is for "contextual conversions to bool". Maybe that can be added with a link to https://en.cppreference.com/w/cpp/language/implicit_conversion#Contextual_conversions?

**Comment by hsutter** [comment]
Editors call: Looks good, thanks!

**Pull request merged**